New Plugin: Heatpump IR (from Playground P115)

TD-er opened the pull request:
A new PR, since I was not able to push commits on the other one.

Closes #2541

ToniA committed:
Add HeatpumpIR 1.0.9 library

ToniA committed:
Add HeatpumpIR plugin P115

TD-er committed:
Rename _P115_HeatpumpIR.ino to _P088_HeatpumpIR.ino

TD-er committed:
Wrap P088 in USES_P088 defines

TD-er committed:
Fix Build issues P088_HeatpumpIR

jimmys01 commented:
The Error:
> deleting object of polymorphic class type 'IRSenderIRremoteESP8266' which has non-virtual destructor might cause undefined behaviour


In my enviroment compiles fine.

ToniA committed:
Fix compile warnings in HeatpumpIR plugin

ToniA commented:
This commit should fix the compile issues. The Travis build has stricter policy on compiler warnings.

https://github.com/ToniA/ESPEasy/commit/9f0bc12514bdc262a20aed907b5cd0a37a253bc7

TD-er commented:
> The Error:
> 
> > deleting object of polymorphic class type 'IRSenderIRremoteESP8266' which has non-virtual destructor might cause undefined behaviour
> 
> In my enviroment compiles fine.

It may compile fine, but it is for sure a bug and I'm very glad it will lead to a compile error.
The thing with non virtual destructors is that if you have a pointer to the base class (normal thing to do when using inheritance) an call the delete on that pointer, what will then be deleted?
It will not call the destructor of the derived class, but only the base class. So there will be at least a memory leak and if stuff has to be done in the destructor, those tasks will not be performed.
This can lead to very hard to reproduce errors.

TD-er commented:
Has anyone tested this?

wixaw commented:
@TD-er  I just saw this new PR, as said on the other conversation, I have the following problem: When I compile, my .pio\build\normal_IRext_ESP8266_4M\firmware.bin is 1022ko and the max of my esp (via the online upload) is 1019ko. I can't send it and i can't test this

ToniA commented:
Yes, I have the same issue here :( Something else should be taken out, or it's not OTA-capable. USB flashing seems to work fine, though.

ToniA commented:
What comes to testing P088_HeatpumpIR, configure a device of type 'Heatpump IR transmitter'. Connect a LED (with resistor like 1k) between the GPIO of your choice and GND. For real use obviously need an IR LED, but for testing you can use any visible color.

For test cases, please refer to the comments in the source code of P088_HeatpumpIR.ino

jimmys01 commented:
@TD-er Perhaps we should consider a minimal build for the 4m build as well?

TD-er commented:
@jimmys01 Yep, one based on the 1M OTA plugin set perhaps?

So one setup with these as primary defines in platformio.ini:

```
platform                  = ${core_2_5_2.platform}
build_flags               = ${core_2_5_2.build_flags} ${common.build_flags} -DPLUGIN_BUILD_MINIMAL_OTA
```

But even for the 4M devices, there is an absolute maximum sketch size of  1044464 bytes.
You should not cross that maximum size.

jimmys01 commented:
Ok some of my thoughts for the 4M build.

Many use cases for the IR plug in are as follows:
Sense high humitidy -> open ac in dry mode or whatever until himitidy drops.
Sense low temp -> open AC in HOT mode until is not cold any more.
Sense PIR not having detect movment fot 20 mins -> Close the AC and TV

To do that without the use of controllers (only a standalone node with rules) you need to have plugins sets in the IR build.
IF we leave the build on OTA then someone that would like to use this functionality would have to compile his own version and that makes the entry to use very high.

So I propose to at least have the extended IR library bundled with the "Enviromental" plug-ins such as temp and humidity sensors.

Any thoughs?

EDIT

```
Checking size .pio\build\normal_IRext_ESP8266_4M\firmware.elf
Memory Usage -> http://bit.ly/pio-memory-usage
DATA:    [======    ]  57.0% (used 46720 bytes from 81920 bytes)
PROGRAM: [==========]  99.7% (used 1041368 bytes from 1044464 bytes)
```
99.7?!  Is that acceptable?

jimmys01 commented:
Proposed changes:
[PR.zip](https://github.com/letscontrolit/ESPEasy/files/3508636/PR.zip)

- Simplified IR defines in platformio.ini and define_plugin_sets.h
- Made the 4M build to be a minimal plug in build as the 1M is

TD-er commented:
> 99.7?! Is that acceptable?

Everything below 100.0 % is acceptable :)  (`<` not `<=`)
Maybe we can also add some more defines to exclude parts of the webserver code base to reduce the binary size.

I will add your zip-of-change to the PR

TD-er committed:
[Heatpump IR] Add minimal IR_ext build for 4M flash nodes

See discussion here: https://github.com/letscontrolit/ESPEasy/pull/2543#issuecomment-521676593

jimmys01 commented:
> Maybe we can also add some more defines to exclude parts of the webserver code base to reduce the binary size.

Lets give it a try because i really like having the normal build with the IR ext functionality.

As you saw 0.3% of free space is olny asking for problems down the road.
Plugins that dont make sence with IR (like led strips?) can be also removed. I dont know of an elegant way to do this thought

TD-er commented:
Just create a define in the `define_plugin_sets.h` file like has been done here:
https://github.com/letscontrolit/ESPEasy/blob/aba4d88ee1b3ed5028cd0d164543c2e1883d9079/src/define_plugin_sets.h#L119-L172
This will allow to make a specific sub set of plugins and controllers for this specific build.
Make sure to set the define you create in the platformio.ini `build_flags`

wixaw commented:
Hello
With minimal_IRext_ESP8266_4M not have sensor BMP :/
Which "USES_XX" do I have use ? 
++

TD-er commented:
> Hello
> With minimal_IRext_ESP8266_4M not have sensor BMP :/
> Which "USES_XX" do I have use ?
> ++

BMP280 is also supported by the BME280 plugin.
I leave this as an exercise to the reader which value to use for the xx ;)
- _P028_BME280.ino .... `#ifdef USES_P028`
- _P030_BMP280.ino ... <fill in the blanks :) >

wixaw commented:
Yessss !
it was the "USES_P006" in my case, but now everything is ok.
Thank you all

loralex commented:
Hi i am testing.

i use the P088.zip file
now the compiling error is ok, and the upload is ok
but in the espeasy device menu the Heatpump IR transmitter  is not showing
i check all the file and any P115 is defined

what i am missing?

loralex commented:
where can i find the last updated Source file where all the fix reported below are included?

maybe i am missing something and the menu Heatpump IR transmitter still missing on my device :(

sorry about, but i am newr here

loralex commented:
Hi all

i just use the  https://github.com/ToniA/ESPEasy  clean without any change

compiling and upload is ok to my Wemos 
looking the memory i am not running out of space as show below:

Sketch uses 863880 bytes (82%) of program storage space. Maximum is 1044464 bytes.
Global variables use 38960 bytes (47%) of dynamic memory, leaving 42960 bytes for local variables. Maximum is 81920 bytes.

BUT still now showing the "Heatpump IR transmitter" from device  menu.
i dont know anymore what to do.

could anyone try to help me?


thx lots

TD-er commented:
It is included in the `PLUGIN_BUILD_IR_EXTENDED`, so you should either build `minimal_ir_extended` or `normal_ir_extended` from PlatformIO.

TD-er commented:
And I made a [test build](https://www.dropbox.com/s/6gshcz4vvmsmtz7/minimal_IRext_ESP8266_1M_4M.rar?dl=0) for you to test this PR.

TD-er committed:
[PlatformIO] Move duplicate lines to global [env]

TD-er committed:
[PlatformIO]  Maintain same order of platform and build_flags

This makes it easier to adapt settings, since platform and build_flags both depend on the same variable section.
Thus it makes sense to have them on consecutive lines.

TD-er committed:
[PlatformIO] Fix missing define on some core 2.4.x builds

ToniA committed:
Configurable MQTT package size in Controller.ino

A problem was found with Domoticz MQTT testing, when sending heatpumpir commands via Domoticz MQTT api

jimmys01 commented:
@TD-er Why on a 4M node we have only 1M available for the program data?

TD-er commented:
> @TD-er Why on a 4M node we have only 1M available for the program data?

That's a legacy of older flash layout definitions in the core libraries.
We still need roughly 2 MB for the sketch if we want to use OTA. (sketch size * 2), so I guess for 4 MB modules a 2/2 layout would be more useful than 3/1.
Problem with changing is, it will format the SPIFFS if it doesn't detect a filesystem at the expected location.
Also it is unclear to me if >1MB SPIFFS will have a speed penalty.
For the 16 MB modules, I tried 2/14 layout, but that was completely unusable with SPIFFS.
Page loading would take seconds.
For 16 MB modules we should go for LittleFS as filesystem and maybe also for all filesystem sizes > 1 MB

TD-er committed:
Remove test_core_260_sdk2_alpha builds

These use too much iram, so cannot be built anymore.

TD-er committed:
[Build] Remove all core 2.6.0 SDK2 builds

They are now obsolete and the test builds no longer fit in the iram.

ToniA commented:
I was wondering if these plugin sets could be added into the definition of PLUGIN_BUILD_MINIMAL_IRext:

- PLUGIN_SET_ONLY_SWITCH
- PLUGIN_SET_ONLY_TEMP_HUM

I have received some questions about having the ability to measure temperature or control switches using rules, directly on the same ESP node which controls the A/C device.

jimmys01 commented:
This is the big question #2590. How to partition the plugins since NORMAL MINIMAL etc don't make sense anymore
EDIT: By the way, also now we have a NORMAL build that only lacks IR RX

TD-er commented:
What @jimmys01 said, please submit some proposed plugin set.
Also we need to reduce the number of binaries for several reasons:
- This is rather confusing and overwhelming now what to choose
- Builds fail due to time limit exceeded.
